# Capstone 1: Data Wrangling (alternate)

### Data Acquisition: Get source data from CSV file

In [1]:
# Import packages and modules
import pandas as pd
import numpy as np
import datetime

In [2]:
# Assign file path
"""
    The original CSV file can be exported from 
    https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj
"""
file_path = r'C:\Users\jwilliams2\Downloads\EMS_Incident_Dispatch_Data.csv'

# Read CSV data into a Pandas dataframe
df = pd.read_csv(file_path,header=0)

df.columns # list all column header names

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (14,15,20) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Index(['CAD_INCIDENT_ID', 'INCIDENT_DATETIME', 'INITIAL_CALL_TYPE',
       'INITIAL_SEVERITY_LEVEL_CODE', 'FINAL_CALL_TYPE',
       'FINAL_SEVERITY_LEVEL_CODE', 'FIRST_ASSIGNMENT_DATETIME',
       'VALID_DISPATCH_RSPNS_TIME_INDC', 'DISPATCH_RESPONSE_SECONDS_QY',
       'FIRST_ACTIVATION_DATETIME', 'FIRST_ON_SCENE_DATETIME',
       'VALID_INCIDENT_RSPNS_TIME_INDC', 'INCIDENT_RESPONSE_SECONDS_QY',
       'INCIDENT_TRAVEL_TM_SECONDS_QY', 'FIRST_TO_HOSP_DATETIME',
       'FIRST_HOSP_ARRIVAL_DATETIME', 'INCIDENT_CLOSE_DATETIME',
       'HELD_INDICATOR', 'INCIDENT_DISPOSITION_CODE', 'BOROUGH', 'ATOM',
       'INCIDENT_DISPATCH_AREA', 'ZIPCODE', 'POLICEPRECINCT',
       'CITYCOUNCILDISTRICT', 'COMMUNITYDISTRICT', 'COMMUNITYSCHOOLDISTRICT',
       'CONGRESSIONALDISTRICT', 'REOPEN_INDICATOR', 'SPECIAL_EVENT_INDICATOR',
       'STANDBY_INDICATOR', 'TRANSFER_INDICATOR'],
      dtype='object')

A downloadable description of each dataset field is available at https://data.cityofnewyork.us/Public-Safety/EMS-Incident-Dispatch-Data/76xm-jjuj in the _Attachments_ section under the file name EMS_incident_dispatch_data_description.xlsx. 

In [3]:
df.shape

(8557848, 32)

In [4]:
df.head()

CAD_INCIDENT_ID       INCIDENT_DATETIME INITIAL_CALL_TYPE  \
0        130010001  01/01/2013 12:00:04 AM            RESPIR   
1        130010002  01/01/2013 12:00:19 AM              CARD   
2        130010004  01/01/2013 12:01:04 AM            ARREST   
3        130010005  01/01/2013 12:01:16 AM              SICK   
4        130010006  01/01/2013 12:01:26 AM            INJURY   

   INITIAL_SEVERITY_LEVEL_CODE FINAL_CALL_TYPE  FINAL_SEVERITY_LEVEL_CODE  \
0                            4          RESPIR                          4   
1                            3            CARD                          3   
2                            1          ARREST                          1   
3                            6            SICK                          6   
4                            5          INJURY                          5   

  FIRST_ASSIGNMENT_DATETIME VALID_DISPATCH_RSPNS_TIME_INDC  \
0    01/01/2013 12:01:45 AM                              Y   
1    01/01/2013 12:01:18 AM                              Y   
2    01/01/2013 12:01:33 AM                              Y   
3    01/01/2013 12:02:12 AM                              Y   
4    01/01/2013 12:01:58 AM                              Y   

   DISPATCH_RESPONSE_SECONDS_QY FIRST_ACTIVATION_DATETIME        ...          \
0                           101    01/01/2013 12:01:51 AM        ...           
1                            59    01/01/2013 12:02:08 AM        ...           
2                            29    01/01/2013 12:01:58 AM        ...           
3                            56    01/01/2013 12:02:55 AM        ...           
4                            32    01/01/2013 12:02:55 AM        ...           

   ZIPCODE POLICEPRECINCT  CITYCOUNCILDISTRICT  COMMUNITYDISTRICT  \
0  10472.0           43.0                 18.0              209.0   
1  10454.0           40.0                  8.0              201.0   
2  11418.0          102.0                 29.0              409.0   
3  10453.0           46.0                 14.0              205.0   
4  10457.0           48.0                 15.0              206.0   

  COMMUNITYSCHOOLDISTRICT CONGRESSIONALDISTRICT REOPEN_INDICATOR  \
0                    12.0                  15.0                N   
1                     7.0                  15.0                N   
2                    27.0                   5.0                N   
3                    10.0                  15.0                N   
4                    10.0                  15.0                N   

  SPECIAL_EVENT_INDICATOR  STANDBY_INDICATOR TRANSFER_INDICATOR  
0                       N                  N                  N  
1                       N                  N                  N  
2                       N                  N                  N  
3                       N                  N                  N  
4                       N                  N                  N  

[5 rows x 32 columns]

In [5]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8557848 entries, 0 to 8557847
Data columns (total 32 columns):
CAD_INCIDENT_ID                   int64
INCIDENT_DATETIME                 object
INITIAL_CALL_TYPE                 object
INITIAL_SEVERITY_LEVEL_CODE       int64
FINAL_CALL_TYPE                   object
FINAL_SEVERITY_LEVEL_CODE         int64
FIRST_ASSIGNMENT_DATETIME         object
VALID_DISPATCH_RSPNS_TIME_INDC    object
DISPATCH_RESPONSE_SECONDS_QY      int64
FIRST_ACTIVATION_DATETIME         object
FIRST_ON_SCENE_DATETIME           object
VALID_INCIDENT_RSPNS_TIME_INDC    object
INCIDENT_RESPONSE_SECONDS_QY      float64
INCIDENT_TRAVEL_TM_SECONDS_QY     float64
FIRST_TO_HOSP_DATETIME            object
FIRST_HOSP_ARRIVAL_DATETIME       object
INCIDENT_CLOSE_DATETIME           object
HELD_INDICATOR                    object
INCIDENT_DISPOSITION_CODE         float64
BOROUGH                           object
ATOM                              object
INCIDENT_DISPATCH_AREA     

In [6]:
# Convert all column header names to lowercase
col_names = [str(name).lower() for name in list(df.columns)]
df.columns = col_names
df.columns

Index(['cad_incident_id', 'incident_datetime', 'initial_call_type',
       'initial_severity_level_code', 'final_call_type',
       'final_severity_level_code', 'first_assignment_datetime',
       'valid_dispatch_rspns_time_indc', 'dispatch_response_seconds_qy',
       'first_activation_datetime', 'first_on_scene_datetime',
       'valid_incident_rspns_time_indc', 'incident_response_seconds_qy',
       'incident_travel_tm_seconds_qy', 'first_to_hosp_datetime',
       'first_hosp_arrival_datetime', 'incident_close_datetime',
       'held_indicator', 'incident_disposition_code', 'borough', 'atom',
       'incident_dispatch_area', 'zipcode', 'policeprecinct',
       'citycouncildistrict', 'communitydistrict', 'communityschooldistrict',
       'congressionaldistrict', 'reopen_indicator', 'special_event_indicator',
       'standby_indicator', 'transfer_indicator'],
      dtype='object')

### Data Cleansing: Define necessary functions

In [7]:
# Define function to drop immaterial rows and columns from dataframe object
def drop_from_df(dfObj):
    """
    Remove columns from dataframe object 
    that are not required for analysis
    """
    # Drop all rows with missing value for 'incident_disposition_code'
    dfObj.dropna(subset=['incident_disposition_code'],inplace=True)
    
    # Identify all columns that contain incident indicator data
    list_of_indicator_cols = [name for name in list(dfObj.columns) 
                              if 'indicator' in str(name).lower() and name !='held_indicator']
    
    # Drop all rows that pertain to outlier incidents
    for name in list_of_indicator_cols:
        index_names = dfObj[dfObj[name]=='Y'].index
        dfObj.drop(index_names, inplace=True)
    
    # Remove columns that contain incident indicator data
    dfObj.drop(list_of_indicator_cols,axis=1,inplace=True)
    
    # Identify and remove all columns that contain district or precinct data (geographic zones)
    list_of_zone_cols = [name for name in list(dfObj.columns) 
                         if ('district' in str(name).lower() or name=='policeprecinct')]
    dfObj.drop(list_of_zone_cols,axis=1,inplace=True)
    
    return dfObj

In [8]:
# Define function to modify dtypes of series in dataframe object
def modify_dtype(dfObj):
    """
    Change the dtype of select columns in 
    dataframe object based on its implied value
    """
    # Create list of all columns that contain ISO8601 datetime
    list_of_datetime_cols = [name for name in list(dfObj.columns) 
                             if 'datetime' in str(name).lower()]

    # Convert dtypes for each element in list to datetime
    for name in list_of_datetime_cols:
        dfObj[name]=pd.to_datetime(dfObj[name],errors='coerce')
       
    # Create list of all columns that contain time duration
    list_of_numeric_cols = [name for name in list(dfObj.columns) 
                             if (('seconds' in str(name).lower()) |
                                 ('severity' in str(name).lower()) |
                                 ('cad' in str(name).lower()))]

    # Convert dtypes for each element in list to numeric
    for name in list_of_numeric_cols:
        dfObj[name]=pd.to_numeric(dfObj[name],errors='coerce')
        
    # Convert columns to category dtypes to reduce size of dataframe object
    dfObj['borough'] = dfObj.borough.astype('category')
    dfObj['zipcode'] = dfObj.zipcode.astype('category')
    dfObj['held_indicator'] = dfObj.held_indicator.astype('category')
    dfObj['valid_dispatch_rspns_time_indc'] = dfObj.valid_dispatch_rspns_time_indc.astype('category')
    dfObj['valid_incident_rspns_time_indc'] = dfObj.valid_incident_rspns_time_indc.astype('category')
    dfObj['incident_dispatch_area'] = dfObj.incident_dispatch_area.astype('category')
    dfObj['incident_disposition_code'] = dfObj.incident_disposition_code.astype('category')
        
    return dfObj

In [9]:
def redesign_df(dfObj):
    """
    Restructure dataframe object
    """
    # CREATE PANDAS SERIES FOR TARGET VARIABLE: fatality
    dfObj['fatality'] = np.logical_or(dfObj.incident_disposition_code == 83,\
                                   dfObj.incident_disposition_code == 96)

    # Create separate columns for the year and month of the incident
    dfObj['incident_year'] = pd.DatetimeIndex(dfObj.incident_datetime).year
    dfObj['incident_month'] = pd.DatetimeIndex(dfObj.incident_datetime).month

    # Reorder dataframe columns
    col_order = ['incident_year','incident_month','cad_incident_id',\
                 'incident_datetime','borough','zipcode',\
                 'incident_dispatch_area','held_indicator',\
                 'initial_call_type','initial_severity_level_code',\
                 'final_call_type','final_severity_level_code',\
                 'first_assignment_datetime','valid_dispatch_rspns_time_indc',\
                 'dispatch_response_seconds_qy','first_activation_datetime',\
                 'first_on_scene_datetime','valid_incident_rspns_time_indc',\
                 'incident_response_seconds_qy','incident_travel_tm_seconds_qy',\
                 'first_to_hosp_datetime','first_hosp_arrival_datetime',\
                 'incident_close_datetime',
                 'incident_disposition_code','fatality']
    dfObj=dfObj[col_order]
    
    return dfObj

### Data Cleansing: Apply functions to dataset

In [10]:
# Apply drop_from_df to dataframe object
df = drop_from_df(df)

In [11]:
df.head()

cad_incident_id       incident_datetime initial_call_type  \
0        130010001  01/01/2013 12:00:04 AM            RESPIR   
1        130010002  01/01/2013 12:00:19 AM              CARD   
2        130010004  01/01/2013 12:01:04 AM            ARREST   
3        130010005  01/01/2013 12:01:16 AM              SICK   
4        130010006  01/01/2013 12:01:26 AM            INJURY   

   initial_severity_level_code final_call_type  final_severity_level_code  \
0                            4          RESPIR                          4   
1                            3            CARD                          3   
2                            1          ARREST                          1   
3                            6            SICK                          6   
4                            5          INJURY                          5   

  first_assignment_datetime valid_dispatch_rspns_time_indc  \
0    01/01/2013 12:01:45 AM                              Y   
1    01/01/2013 12:01:18 AM                              Y   
2    01/01/2013 12:01:33 AM                              Y   
3    01/01/2013 12:02:12 AM                              Y   
4    01/01/2013 12:01:58 AM                              Y   

   dispatch_response_seconds_qy first_activation_datetime   ...     \
0                           101    01/01/2013 12:01:51 AM   ...      
1                            59    01/01/2013 12:02:08 AM   ...      
2                            29    01/01/2013 12:01:58 AM   ...      
3                            56    01/01/2013 12:02:55 AM   ...      
4                            32    01/01/2013 12:02:55 AM   ...      

  incident_travel_tm_seconds_qy  first_to_hosp_datetime  \
0                         696.0  01/01/2013 12:28:49 AM   
1                         792.0                     NaN   
2                         400.0                     NaN   
3                         772.0  01/01/2013 12:34:54 AM   
4                         824.0  01/01/2013 12:27:42 AM   

   first_hosp_arrival_datetime  incident_close_datetime held_indicator  \
0       01/01/2013 12:38:15 AM   01/01/2013 01:04:56 AM              N   
1                          NaN   01/01/2013 12:55:34 AM              N   
2                          NaN   01/01/2013 12:38:05 AM              N   
3       01/01/2013 12:53:02 AM   01/01/2013 01:20:28 AM              N   
4       01/01/2013 12:31:13 AM   01/01/2013 12:53:12 AM              N   

  incident_disposition_code borough atom  incident_dispatch_area  zipcode  
0                      82.0   BRONX  NaN                      B3  10472.0  
1                      93.0   BRONX  NaN                      B1  10454.0  
2                      83.0  QUEENS  NaN                      Q3  11418.0  
3                      82.0   BRONX  NaN                      B2  10453.0  
4                      82.0   BRONX  NaN                      B2  10457.0  

[5 rows x 23 columns]

In [12]:
# Apply modify_dtype to dataframe object
df = modify_dtype(df)

In [13]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8431649 entries, 0 to 8546304
Data columns (total 23 columns):
cad_incident_id                   int64
incident_datetime                 datetime64[ns]
initial_call_type                 object
initial_severity_level_code       int64
final_call_type                   object
final_severity_level_code         int64
first_assignment_datetime         datetime64[ns]
valid_dispatch_rspns_time_indc    category
dispatch_response_seconds_qy      int64
first_activation_datetime         datetime64[ns]
first_on_scene_datetime           datetime64[ns]
valid_incident_rspns_time_indc    category
incident_response_seconds_qy      float64
incident_travel_tm_seconds_qy     float64
first_to_hosp_datetime            datetime64[ns]
first_hosp_arrival_datetime       datetime64[ns]
incident_close_datetime           datetime64[ns]
held_indicator                    category
incident_disposition_code         category
borough                           category
ato

In [14]:
# Apply redesign_df to dataframe object
df = redesign_df(df)

In [15]:
# Create a multindex on 'incident_year', 'incident_month' and 'cad_incident_id'
df.set_index(['incident_year','incident_month','cad_incident_id'])

incident_datetime  \
incident_year incident_month cad_incident_id                       
2013          1              130010001       2013-01-01 00:00:04   
                             130010002       2013-01-01 00:00:19   
                             130010004       2013-01-01 00:01:04   
                             130010005       2013-01-01 00:01:16   
                             130010006       2013-01-01 00:01:26   
                             130010007       2013-01-01 00:01:35   
                             130010008       2013-01-01 00:01:52   
                             130010009       2013-01-01 00:02:11   
                             130010010       2013-01-01 00:02:56   
                             130010011       2013-01-01 00:03:12   
                             130010012       2013-01-01 00:03:12   
                             130010013       2013-01-01 00:03:29   
                             130010014       2013-01-01 00:03:41   
                             130010015       2013-01-01 00:03:47   
                             130010016       2013-01-01 00:05:07   
                             130010017       2013-01-01 00:05:08   
                             130010018       2013-01-01 00:05:19   
                             130010019       2013-01-01 00:05:37   
                             130010020       2013-01-01 00:05:45   
                             130010022       2013-01-01 00:05:52   
                             130010023       2013-01-01 00:06:12   
                             130010024       2013-01-01 00:06:30   
                             130010026       2013-01-01 00:07:22   
                             130010027       2013-01-01 00:07:28   
                             130010028       2013-01-01 00:07:31   
                             130010029       2013-01-01 00:07:33   
                             130010030       2013-01-01 00:07:46   
                             130010031       2013-01-01 00:08:22   
                             130010032       2013-01-01 00:08:26   
                             130010034       2013-01-01 00:08:41   
...                                                          ...   
2018          12             183654353       2018-12-31 23:48:39   
                             183654354       2018-12-31 23:48:57   
                             183654356       2018-12-31 23:49:28   
                             183654357       2018-12-31 23:49:41   
                             183654358       2018-12-31 23:49:42   
                             183654359       2018-12-31 23:50:12   
                             183654360       2018-12-31 23:50:38   
                             183654361       2018-12-31 23:50:47   
                             183654362       2018-12-31 23:50:49   
                             183654363       2018-12-31 23:51:00   
                             183654364       2018-12-31 23:51:28   
                             183654365       2018-12-31 23:51:39   
                             183654366       2018-12-31 23:52:03   
                             183654367       2018-12-31 23:52:09   
                             183654369       2018-12-31 23:52:40   
                             183654370       2018-12-31 23:53:23   
                             183654372       2018-12-31 23:53:30   
                             183654373       2018-12-31 23:53:51   
                             183654374       2018-12-31 23:54:02   
                             183654375       2018-12-31 23:55:23   
                             183654376       2018-12-31 23:55:27   
                             183654377       2018-12-31 23:55:49   
                             183654379       2018-12-31 23:57:04   
                             183654380       2018-12-31 23:57:06   
                             183654381       2018-12-31 23:57:41   
                             183654382       2018-12-31 23:57:42   
                             183654

### Inspect dataframe

In [16]:
df.shape

(8431649, 25)

In [17]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8431649 entries, 0 to 8546304
Data columns (total 25 columns):
incident_year                     int64
incident_month                    int64
cad_incident_id                   int64
incident_datetime                 datetime64[ns]
borough                           category
zipcode                           category
incident_dispatch_area            category
held_indicator                    category
initial_call_type                 object
initial_severity_level_code       int64
final_call_type                   object
final_severity_level_code         int64
first_assignment_datetime         datetime64[ns]
valid_dispatch_rspns_time_indc    category
dispatch_response_seconds_qy      int64
first_activation_datetime         datetime64[ns]
first_on_scene_datetime           datetime64[ns]
valid_incident_rspns_time_indc    category
incident_response_seconds_qy      float64
incident_travel_tm_seconds_qy     float64
first_to_hosp_datetime     

### Export dataframe to CSV

In [18]:
# Export dataframe to CSV
output_path = r'C:\Users\jwilliams2\Desktop\cp1\clean_comp_df.csv'
print('Exporting dataframe to CSV...')
df.to_csv(output_path,index=False,compression='gzip')
print('Dataframe successfully exported to CSV using \'gzip\' compression.')

Exporting dataframe to CSV...
Dataframe successfully exported to CSV using 'gzip' compression.


###### Summary
The original dataset was comprised of 8,557,848 observations of mixed data types, and its source file occupied approximately 1.98 GB of hard disk space and 10.9 GB in system memory. The target variable ("fatality") was created by applying a boolean filter on the "incident_disposition_code" column within the dataframe, which indicates the outcome of any EMS incident. 

After all data pre-processing was complete, the resulting clean dataset was comprised of 8,431,649 observations of mixed data types, with a clear target variable and 21 predictor variables. Its output file occupied 299 MB of hard disk space and 2.0 GB in system memory.